pkg/daemon/daemon.go

```diff
@@ -398,54 +435,65 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
 		// Apply generic_plugin last
 		err = dn.LoadedPlugins[GenericPlugin].Apply()
 		if err != nil {
-			glog.Errorf("nodeStateChangeHandler(): generic_plugin fail to apply: %v", err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): generic_plugin fail to apply: %v", err)
+			return err
 		}
 	}
 
 	if reqReboot {
-		glog.Info("nodeStateChangeHandler(): reboot node")
-		go rebootNode()
-		return
+		glog.Info("nodeStateSyncHandler(): reboot node")
+		rebootNode()
+		return nil
 	}
 
 	// restart device plugin pod
-	if reqDrain || (latestState.Spec.DpConfigVersion != oldState.Spec.DpConfigVersion) {
-		glog.Info("nodeStateChangeHandler(): restart device plugin pod")
+	if reqDrain || (dn.nodeState.GetName() != "" && latestState.Spec.DpConfigVersion != dn.nodeState.Spec.DpConfigVersion) {
+		glog.Info("nodeStateSyncHandler(): restart device plugin pod")
 		if err := dn.restartDevicePluginPod(); err != nil {
-			glog.Errorf("nodeStateChangeHandler(): fail to restart device plugin pod: %v", err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): fail to restart device plugin pod: %v", err)
+			return err
 		}
 	}
 
 	if anno, ok := dn.node.Annotations[annoKey]; ok && anno == annoDraining {
-		dn.completeDrain()
+		if err := dn.completeDrain(); err != nil {
+			glog.Errorf("drainNode(): failed to complete draining: %v", err)
+			return err
+		}
 	} else if !ok {
 		if err := dn.annotateNode(dn.name, annoIdle); err != nil {
 			glog.Errorf("drainNode(): failed to annotate node: %v", err)
-			dn.exitCh <- err
-			return
+			return err
 		}
 	}
+	glog.Info("nodeStateSyncHandler(): sync succeeded")
+	dn.nodeState = latestState.DeepCopy()
 	dn.refreshCh <- Message{
 		syncStatus:    "Succeeded",
 		lastSyncError: "",
 	}
+	// wait for writer to refresh the status
+	<-dn.syncCh
```

Comment:
This is for the scenario of https://bugzilla.redhat.com/show_bug.cgi?id=1815039

As in the `nodeStateSyncHandler`, we only check the generation of the CR. However, the generation can only reflect the change in the `spec` of CR. Without this sync, if users re-apply a policy right after deleting it, there is a change that the spec and status in the `latestState` are all in the desired state, as the deletion is not finished, the status is yet up-to-date. It will make the `needUpdate` function will return `false`. 

This sync blocks the change handler until the writer update the status. So that the latest state in invoking will not has such wrong information.

I don't think we need this sync when setting status to `InProgress`, as `nodeStateSyncHandler` doesn't exit at that time. But I think we may need it for the `Failed` case as well.

Reply:
ack